### PR TITLE
Replace kernel builder with constructor

### PR DESCRIPTION
**Pull request by johanste**
### Motivation and Context

> Attempt to make the code slightly more pythonic (work in progress).

This PR removes the kernel.builder() pattern ported from the C# semantic kernel. As part of this change, a kernel cannot be built with memory_storage because an AI service must be registered after creation. Note it still can created with a prebuilt SemanticTextMemory (contains an embedding generation service). This can be improved upon in a future PR. 

The new pattern is now:

`kernel = sk.Kernel()`

If a user wants to use memories, they can do:

`kernel.add_embedding_backend()`
`kernel.register_memory_storage()`

### Description

The builder pattern (including the chained configuration) felt a bit out of place for me as a python developer. This change removes the kernel builder and instead use keyword-only parameters in the constructor of the kernel.

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/kernel.py [hunk 1]
```diff
@@ -43,17 +49,22 @@ class Kernel(KernelBase):
...
-        self._skill_collection = skill_collection
-        self._prompt_template_engine = prompt_template_engine
-        self._memory = memory
+        self._log = log if log else NullLogger()
+        self._config = config if config else KernelConfig()
+        self._skill_collection = skill_collection if skill_collection else SkillCollection(self._log)
+        self._prompt_template_engine = prompt_template_engine if prompt_template_engine else PromptTemplateEngine(self._log)
+        self._memory = memory if memory else NullMemory()
+
+        if memory_storage:
```
Comment (dluc) [inline python/semantic_kernel/kernel.py:66]: this is one of those cases why the builder is a better path. With a growing set of dependencies, this approach turns the kernel ctor into a big builder logic. The builder allows to move the composition out, so for example the kernel doesn't need to know that memory internally uses memory_storage

**Review by lukins — CHANGES_REQUESTED** [review]
Skvělé

**Comment by jjhenkel** [comment]
I want to chime in here and say, after some discussion, we are really looking to head toward making the Python version more idiomatic. (For context, I did a direct port from C# initially to kick-start things and now we'd love to make it more natural/pythonic.) 

I fully support removing the builder-style Kernel creation. The C# port brings over many layers of indirection that really don't feel necessary for Python.

File: python/semantic_kernel/kernel.py (hunk 1 above)
Comment (AdityaGudimella) [reply python/semantic_kernel/kernel.py:66]: One suggestion here is to convert the class into a dataclass. That way, all the `x if x else NullX` can be converted to `x: XType = field(default_factory=NullX)`.
```suggestion
       @dataclass
       class Kernel(KernelBase):
           _log: Logger = field(default_factory=NullLogger)
           _config: field(default_factory=KernelConfig)
```

One downside of this is that if you want to instantiate the class and call the constructor using keyword arguments, the keyword arguments will be starting with `_`. This is one of the reasons why I prefer `attrs` to dataclasses. With `attrs` the above definition would automatically translate to keyword arguments without underscores, but the attributes will automatically be considered "private".

Comment (johanste) [reply python/semantic_kernel/kernel.py:66]: Do we have to make the attributes private?

**Comment by dluc** [comment]
There are several non-idiomatic aspect of this initial alpha, and that's by design, to get the lib functional. There's going to be a second stage where we focus on code quality. This is also a particular one that I would leave it last, since it touches also the "design", so we could be spending days debating :-)

File: python/semantic_kernel/kernel.py (hunk 1 above)
Comment (AdityaGudimella) [reply python/semantic_kernel/kernel.py:66]: I'm not opposed to making them public.

**Comment by dluc** [comment]
@jjhenkel @AdityaGudimella ok to move forward with this, there's a couple of conflicts to fix. Please let me know how you prefer to proceed

**Comment by dluc** [comment]
@johanste would you like to update fixing the conflicts so we would merge?

**Comment by johanste** [comment]
I fixed the conflicts, but I am not comfortable with how much (or rather little) testing I have done. Unfortunately, I'm going offline for two weeks starting... now... If anyone wants to take over this, it would be great; else I will pick it up again when I'm back.

**Comment by dluc** [comment]
@mkarle @awharrison-28 could you review and test please?

**Comment by awharrison-28** [comment]
Reviewing and testing

**Comment by shawncal** [comment]
FYI: Python PRs on hold while we stage to merge /python into main. See https://github.com/microsoft/semantic-kernel/pull/423

**Comment by awharrison-28** [comment]
@mkarle @dluc @johanste I took over development of the PR. 

I updated documentation, fixed unit tests, and verified functionality of the notebooks on my local copy.

**Review by alexchaomander — APPROVED** [review]
LGTM!

File: samples/notebooks/python/1-basic-loading-the-kernel.ipynb [hunk 2]
```diff
@@ -168,7 +160,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.11.3"
+   "version": "3.10.10"
```
Comment (mkarle) [inline samples/notebooks/python/1-basic-loading-the-kernel.ipynb:163]: Why drop the python version?

**Review by mkarle — APPROVED** [review]
Ran the tests and end-to-end test files. LGTM

**Comment by dluc** [comment]
@johanste we've merged the python branch into `main`, and GitHub doesn't allow to point this PR to `main` because of some rebase steps (I tried and GH automatically closed the PR, without an option to reopen it). 

Could you send the PR again?

**Pull request closed without merging**